Game logic

sumancrest0001 opened the pull request.

sumancrest0001 committed:
players class added

Commit:
working on UI

sumancrest0001 committed:
ready for second milestone

Commit:
lint fixes

Commit:
fixed lint errors

sumancrest0001 committed:
game logic mock up

Commit:
board display

Commit:
second milestone

sumancrest0001 committed:
interconnection between classes is remaining

sumancrest0001 committed:
user-interface is prepared

sumancrest0001 committed:
user-interface is edited

sumancrest0001 committed:
user-interface, linter error fixed

sumancrest0001 committed:
user-interface, linter error fixed final:

sumancrest0001 committed:
made the changes

sumancrest0001 committed:
linter errors fixed

sumancrest0001 committed:
linter errors resolved

sumancrest0001 committed:
changes done

sumancrest0001 committed:
linter errors fixed

sumancrest0001 committed:
fixed linter errors

sumancrest0001 committed:
fixed linter error

sumancrest0001 committed:
made all the changes

sumancrest0001 committed:
random

sumancrest0001 committed:
finalized the game and its working

sumancrest0001 committed:
solved the linter errors

sumancrest0001 committed:
linter errors fixed

bolah2009 commented on bin/main.rb:
```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
```
Let's make `main.rb` an executable file.

Adding `#!/usr/bin/env ruby` to this line will make `bin/main.rb` run as `ruby bin/main.rb` would.
```suggestion
#!/usr/bin/env ruby
# frozen_string_literal: true
```

bolah2009 commented on bin/main.rb:
```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+require_relative '../lib/game_logic'
+class UserInterface
```
Please let's move this class to the `lib` directory.

> All classes are kept in `lib` directory.

bolah2009 commented on lib/game_logic.rb:
```diff
@@ -0,0 +1,51 @@
+
+class GameLogic
+  attr_reader :player1_array, :player2_array, :wining_compositions
+  attr_accessor :winner, :cells, :player1, :player2, :current_player, :mark
```
Why do we need to change the value of `winner` from outside the `Class`? :thinking: 

Don't you think allowing writing of this property from out the class is unnecessary, by doing this the variable can be set to anything? 

Let's keep all variables that don't necessarily need to be changed from outside the Class as `attr_reader`.


> Use `attr_readers`/`attr_writers `instead of `atrr_acessors`.

bolah2009 commented on lib/game_logic.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  def check_winner
+    for i in 0...8
```
You can use `each` to fix linter warning/error.

```suggestion
    (0...8).each do |i|
```

bolah2009 commented on lib/game_logic.rb:
```diff
@@ -0,0 +1,51 @@
+    @cells = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+  end
+
+  def check_winner
```
:bulb: You can keep must of these methods private by having a method e.g. `play_game` that checks for a winner, updates the cell, etc from inside the class. That way you are not exposing too many methods and variable from outside the class.

bolah2009 reviewed (changes requested):
# Hi @sumancrest0001  

### Awesome job with [the project](https://github.com/sumancrest0001/ruby-tic-tac-toe/tree/ab5af58fac0728b60725026b6b0d66adbc6ec4b4). :clap:  Almost there :smile: 

### The game flow works perfectly, handle payers valid and invalid inputs correctly, checks for winner and draw accurately. :100: :clap: 

A few things to change before moving to the next (and **final**) milestone :blush:  :
 - Kindly rename the PR (to the name of the milestone) and the PR description to a brief summary of what was done in this PR. (Examples Below)
- Fix linter errors/warnings
- All classes are kept in `lib` directory.
- Use `attr_readers`/`attr_writers` instead of `atrr_acessors`.
- See other comments for required changes.

Kindly, check the comment below for suggestions.

#### :recycle:  Make these changes and submit another code review request. :+1: 

#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

bolah2009 reviewed (commented):
# Rename PR

  ![rename PR](https://user-images.githubusercontent.com/36057474/63838245-ec569f80-c974-11e9-93a9-83a4db5c259f.gif)

-----

# Update PR description

  ![update-PR-description](https://user-images.githubusercontent.com/36057474/64303100-8672a500-cf7e-11e9-999c-d23c06358f0e.gif)

sumancrest0001 committed:
made all the required changes

sumancrest0001 commented:
Milestone-3
1. made changes on the accessibility of instance variables.
2. made the main file executable only.

Torres-ssf commented:
### Pull Request Under Review By Sérgio (TSE)

Torres-ssf commented on lib/game_logic.rb:
```diff
@@ -0,0 +1,169 @@
+  # this code gets where the user marks
+  def choose_mark
+    loop do
+      puts "Player, #{game.current_player}. Please choose a box that you want to mark"
```
**[REQUIRED]**
We need to give a feedback message for the players when an invalid input is sent. Something like "Invalid input, chose a number between 1-9.

**[REQUIRED]**
You can also give a feedback when one player tries to move to a position already taken.

Torres-ssf commented on lib/game_logic.rb:
```diff
@@ -0,0 +1,169 @@
+  def initialize
+    @wining_compositions = [[1, 2, 3], [4, 5, 6], [7, 8, 9], [1, 4, 7], [2, 5, 8], [3, 6, 9], [1, 5, 9], [3, 5, 7]]
+    @winner = nil
+    @player1 = nil
```
**[OPTIONAL]**
Create a Player class and let the player class handle everything related to players. This will make your code follow even more the OOP principles, adding high value to it. So each player object is going to have its own mark and other objects that you may want to add in the future, like name.

Torres-ssf commented on lib/game_logic.rb:
```diff
@@ -0,0 +1,169 @@
+    @current_player = nil
+    @player1_array = []
+    @player2_array = []
+    @cells = [1, 2, 3, 4, 5, 6, 7, 8, 9]
```
**[OPTIONAL]**
Create a Board class and let the board class handle everything related to the board. Then create methods for the board check for empty cells, update cells. This will also make your code more OOP.

Torres-ssf reviewed (changes requested):
### Nicely done team :clap: 

### Congratulations, your game logic is great and your code is clean :clap: :tada: 

But we still have some minor changes before your :+1: 

I left the required suggestion and some optional ones that will make your code even better.

Please make these changes and submit for code review again.
Happy coding :computer:

sumancrest0001 committed:
made the required changes

sumancrest0001 committed:
linter error fixed

sumancrest0001 committed:
linter errors solved

sumancrest0001 committed:
linter solved

sumancrest0001 commented:
milestone-3
1. added instructions for the players when their marks are invalid

maelfosso commented on lib/game_logic.rb:
```diff
@@ -0,0 +1,176 @@
+    @game.player2 = @game.player1 == 'X' ? 'O' : 'X'
+    puts "Player1, Your Symbol is #{game.player1} and Player2, your symbol is #{game.player2}"
+    @game.current_player = @game.player1
+  end
```
What if you initialize the player into the GameLogic class ?
```ruby
def initialize(player1, player2)
  @player1 = Player.new(player1)
  @player2 = Player.new(player2)
  @current_player = ....
end
```

maelfosso commented on lib/game_logic.rb:
```diff
@@ -0,0 +1,176 @@
+        puts 'This is an invalid input Please select one of the rremaining cells number'
+      elsif (@game.cells.include? @game.mark) == false
+        puts 'This cell is already marked please select new one'
+      end
```
Please put all those 6 previous lines into a method. Something like
```ruby
def valid_input?(input)
    return false if (mark > 9 || mark < 1) || @game.cells.include? input
    return true
end
```
That is because is not good to manipulate class attribute outside of his class like you do.

maelfosso reviewed (changes requested):
Hello coders :smile: 

It was a pleasure to read you're code. 
Thank for the improvements :clap: 
You're some steps before closing the feature. Please read what you've to do below.

- [ ] **No puts, gets.chomp, print into any class in `lib/` folder**

- [ ] **One file for only and only one class** even if Ruby allow us to have many classes into one file.

- [ ] Move the **UserInterface** class into `bin/main.rb` 

- [ ] **You must have 2 classes into lib/ folder**. What if you create a class for `Player`. Take a look on comments below to know how you could refactor your code.

Kindly make those changes and submit a new code review request.
Thanks.

sumancrest0001 committed:
made the changes

sumancrest0001 committed:
linter errors fixed

sumancrest0001 commented:
Milestone-3
1. Added another class in the lib folder.
2. Moved the user_interface class in the main.rb file
3. Defined separate method to check invalid inputs.

samgaco commented on bin/main.rb:
```diff
@@ -0,0 +1,129 @@
+  def play_game
+    display_instructions
+    set_players
+    until @game.game_end(@player.winner)
```
**[OPTIONAL]** this could be in the initializer (some people would place the UI logic outside of the class, but it is also fine right here), that way we keep all the methods in our class with one single responsibility




----------------



> Methods, like classes, should have a single responsibility. All of the same reasons
> apply; having just one responsibility makes them easy to change and easy to reuse. All
> the same design techniques work; ask them questions about what they do and try to
> describe their responsibilities in a single sentence. 

> 
>  Methods that have a single responsibility confer the following benefits:
> • **Expose previously hidden qualities** Refactoring a class so that all of its methods
> have a single responsibility has a clarifying effect on the class. Even if you do not
> intend to reorganize the methods into other classes today, having each of them
> serve a single purpose makes the set of things the class does more obvious.
> • **Avoid the need for comments** How many times have you seen a comment that is
> out of date? Because comments are not executable, they are merely a form of
> decaying documentation. If a bit of code inside a method needs a comment,
> extract that bit into a separate method. The new method name serves the same
> purpose as did the old comment.
> • **Encourage reuse** Small methods encourage coding behavior that is healthy for
> your application. Other programmers will reuse the methods instead of duplicating the code. They will follow the pattern you have established and create small,
> reusable methods in turn. This coding style propagates itself.
> • **Are easy to move to another class** When you get more design information and
> decide to make changes, small methods are easy to move. You can rearrange behavior
> without doing a lot of method extraction and refactoring. Small methods lower
> the barriers to improving your design.


_**ref: Practical Object-Oriented Design in Ruby: An Agile Primer by Sandi Metz**_

samgaco reviewed (changes requested):
Hi @sumancrest0001 and @andynarf :high_brightness: 


* Right now the game is not displaying the board after asking for the symbols, the game should display the board or at least indicate which moves are available. 
* The linter failing checks should be solved.

Submit back for another code review once you get this done :+1: 

--------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

sumancrest0001 committed:
linter error fixed

sumancrest0001 committed:
magic comment is added

sumancrest0001 committed:
last error fixed

sumancrest0001 committed:
made the changes

bolah2009 reviewed (approved):
It's a :+1: from me.

The game works well :clap: 

You can merge this and proceed to milestone 4.
